### PR TITLE
pg_dump: try to omit expensive external partition queries

### PR DESCRIPTION
**Pull request by jchampio**
The query on the `pg_partitions` view to find external partitions is expensive and currently runs for every partitioned table. Optimization of the view itself has proven tricky -- many corner cases need to be evaluated -- but it seems pointless to run the query if we know we don't have any external partitions to begin with. This is especially true for pg_upgrade, which currently disallows external partitions entirely.

Add a (much faster) preliminary query on `pg_partition[_rule]` to see if the current table has any external partitions, and if not, skip the expensive query entirely.

An unscientific performance comparison: before, it took roughly four minutes to dump the ICW cluster schema on CI hardware; now it takes less than a minute and a half.

### CODE REVIEW

**Comment by ivannovick** [comment]
again nice to see these optimizations :)

File: src/bin/pg_dump/pg_dump.c [hunk 1]
```diff
@@ -14264,6 +14263,21 @@ dumpTableSchema(Archive *fout, TableInfo *tbinfo)
 
 				PQclear(res);
 			}
+
+			/* Find out if there are any external partitions. */
+			resetPQExpBuffer(query);
+			appendPQExpBuffer(query, "SELECT COUNT(*) > 0 AS has_external_partitions "
```
Comment (hlinnaka) [inline src/bin/pg_dump/pg_dump.c]: I'd suggest `SELECT 1 FROM ... LIMIT 1` instead of `COUNT(*) > 0`. It potentially gives a cheaper plan.

Comment (hlinnaka) [reply src/bin/pg_dump/pg_dump.c]: If I'm reading this correctly, you're running this even if we figured out above the the table isn't partitioned at all (the former isPartitioned flag). Skipping this if the table is not partitioned at all still seems like a good optimization.

Any chance this query could be merged with the `SELECT "pg_get_partition_def('%u'::pg_catalog.oid, true)` query above?

Comment (jchampio) [reply src/bin/pg_dump/pg_dump.c]: My first attempt (taking into account your previous comment too) is something like
```
SELECT pg_get_partition_def('%u'::pg_catalog.oid, true),
       EXISTS (
         SELECT 1
           FROM pg_class part
           JOIN pg_partition_rule pr ON (part.oid = pr.parchildrelid)
           JOIN pg_partition p ON (pr.paroid = p.oid)
          WHERE p.parrelid = '%u'::pg_catalog.oid
            AND part.relstorage = '%c'
       ) AS has_external_partitions;
```
It does have the downside that both sides run even if there's no partitioning scheme on the table, though; do you have a gut feeling for which is more important?

**Comment by hlinnaka** [comment]
On 10/10/2018 19:00, Jacob Champion wrote:
> *@jchampio* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/bin/pg_dump/pg_dump.c 
> <https://github.com/greenplum-db/gpdb/pull/5957#discussion_r224140705>:
> 
>> @@ -14264,6 +14263,21 @@ dumpTableSchema(Archive *fout, TableInfo *tbinfo)
>   
>   				PQclear(res);
>   			}
> +
> +			/* Find out if there are any external partitions. */
> +			resetPQExpBuffer(query);
> +			appendPQExpBuffer(query, "SELECT COUNT(*) > 0 AS has_external_partitions "
> 
> My first attempt (taking into account your previous comment too) is 
> something like
> 
> |SELECT pg_get_partition_def('%u'::pg_catalog.oid, true), EXISTS ( 
> SELECT 1 FROM pg_class part JOIN pg_partition_rule pr ON (part.oid = 
> pr.parchildrelid) JOIN pg_partition p ON (pr.paroid = p.oid) WHERE 
> p.parrelid = '%u'::pg_catalog.oid AND part.relstorage = '%c' ) AS 
> has_external_partitions; |
> 
> It does have the downside that both sides run even if there's no 
> partitioning scheme on the table, though; do you have a gut feeling for 
> which is more important?

Hmm. I'm inclined to do a separate query, after all. As you said, the 
server can't do anything smart with the above, it'll just run both parts 
one after another. Submitting them in one query saves one round-trip, 
but I don't think that's so important here. Separate queries seems more 
readable, too.

- Heikki

**Comment by jchampio** [comment]
@hlinnaka Updated with your feedback.

**Comment by doty-pivotal** [comment]
Took a look, nothing jumps out at me.

**Comment by jchampio** [comment]
Thanks for the reviews!

**Pull request merged**